### PR TITLE
🧹 [Clean] Add cache categories argument

### PR DESCRIPTION
**Pull request by danyf90**
Resolves https://github.com/tuist/tuist/issues/3402

### Short description 📝

Adds optional cache category argument to `tuist clean`. If not passed, the behavior is the same as current command.
```
Examples:
tuist clean -> clean everything
tuist clean builds -> clean only build artifacts
tuist clean builds tests-> clean build and test artifacts
```

### Checklist ✅

- [ ] The code architecture and patterns are consistent with the rest of the codebase.
- [ ] The changes have been tested following the [documented guidelines](https://docs.tuist.io/contributors/testing-strategy/).
- [ ] The `CHANGELOG.md` has been updated to reflect the changes. In case of a breaking change, it's been flagged as such.
- [ ] In case the PR introduces changes that affect users, the documentation has been updated.

### CODE REVIEW

File: Sources/TuistCore/Cache/CacheDirectoriesProvider.swift [hunk 1]
```diff
@@ -38,39 +20,26 @@ public final class CacheDirectoriesProvider: CacheDirectoriesProviding {
...
-        ]
+extension CacheCategory {
+    var directoryName: String {
+        switch self {
+        case .plugins:
+            return "Plugins"
+        case .builds:
+            return "BuildCache"
+        case .tests:
+            return "TestsCache"
```
Comment (luispadron) [inline Sources/TuistCore/Cache/CacheDirectoriesProvider.swift:36]: Maybe for 2.0 we could rename these to Builds and Tests, so it matches the other directories?

File: Sources/TuistKit/Commands/CleanCommand.swift [hunk 2]
```diff
@@ -9,7 +10,12 @@ struct CleanCommand: ParsableCommand {
         )
     }
 
+    @Argument(help: "The cache categories to be cleaned. If no category is specified, the whole cache is cleaned.")
```
Comment (luispadron) [inline Sources/TuistKit/Commands/CleanCommand.swift:13]: Is this the desired default? Should we clear everything? Some tools like cocoapods clean just certain parts and then allow specifying clearing everything, just wondering what would be the best here

File: Sources/TuistSupport/HTTP/HTTPRequestDispatcher.swift [hunk 3]
```diff
@@ -77,12 +77,14 @@ public final class HTTPRequestDispatcher: HTTPRequestDispatching {
                 if let error = error {
                     observer(.error(HTTPRequestDispatcherError.urlSessionError(error)))
                 } else if let data = data, let response = response as? HTTPURLResponse {
+                    print("Request \(resource.description), response: \(String(data: data, encoding: .utf8))")
```
Comment (luispadron) [inline Sources/TuistSupport/HTTP/HTTPRequestDispatcher.swift:80]:
```suggestion

```

File: Sources/TuistSupport/HTTP/HTTPRequestDispatcher.swift [hunk 4]
```diff
@@ -77,12 +77,14 @@ public final class HTTPRequestDispatcher: HTTPRequestDispatching {
...
                     switch response.statusCode {
                     case 200 ..< 300:
                         do {
                             let object = try resource.parse(data, response)
                             observer(.success((object: object, response: response)))
                         } catch {
+                            print(error)
```
Comment (luispadron) [inline Sources/TuistSupport/HTTP/HTTPRequestDispatcher.swift:87]:
```suggestion
```

File: Sources/TuistKit/Commands/CleanCommand.swift (hunk 2 above)
Comment (danyf90) [reply Sources/TuistKit/Commands/CleanCommand.swift:13]: For sure should be like this until 2.0

File: Sources/TuistCore/Cache/CacheDirectoriesProvider.swift (hunk 1 above)
Comment (danyf90) [reply Sources/TuistCore/Cache/CacheDirectoriesProvider.swift:36]: Yep, I agree it would makes sense, let's see what @tuist/core thinks

**Review by luispadron — APPROVED** [review]
Looks good, would love to see some of those other breaking changes discussed for 2.0 after this merges.

File: Sources/TuistCore/Cache/CacheDirectoriesProvider.swift (hunk 1 above)
Comment (fortmarek) [reply Sources/TuistCore/Cache/CacheDirectoriesProvider.swift:36]: I think that makes sense as well 👍

File: Sources/TuistKit/Commands/CleanCommand.swift (hunk 2 above)
Comment (fortmarek) [reply Sources/TuistKit/Commands/CleanCommand.swift:13]: as a default, I'd probably clean everything related to the current project. We'd have to restructure a bit the internal structure to do this efficiently but it imho makes the most sense.

File: Sources/TuistKit/Commands/CleanCommand.swift [hunk 5]
```diff
@@ -9,7 +10,12 @@ struct CleanCommand: ParsableCommand {
...
+    var cacheCategories: [CacheCategory] = CacheCategory.allCases
```
Comment (fortmarek) [inline Sources/TuistKit/Commands/CleanCommand.swift:14]: does the help for this property describes what the individual values user can use and what they actually represent?

**Review by fortmarek — COMMENTED** [review]
Thanks, @danyf90! Another thing to note is that the website documentation does not seem to have changed. After that, I'm ready to approve this ✅

File: Sources/TuistKit/Commands/CleanCommand.swift (hunk 5 above)
Comment (danyf90) [reply Sources/TuistKit/Commands/CleanCommand.swift:14]: yep, the values are automatically added by the library:
```
ARGUMENTS:
  <cache-categories>      The cache categories to be cleaned. If no category is
                          specified, the whole cache is cleaned. (default:
                          plugins, builds, tests, generatedAutomationProjects,
                          projectDescriptionHelpers, manifests)
```

I wouldn't go into the details of what each of them actually cleans

File: projects/docs/docs/commands/clean.md [hunk 6]
```diff
@@ -12,3 +12,9 @@ To clean all the data generated by Tuist, you can use the following command:
 ```
 tuist clean
 ```
+
+To clean only a subset of the data generated by Tuist, you can specify the categories to be deleted, for example:
```
Comment (fortmarek) [inline projects/docs/docs/commands/clean.md:16]: I think the website documentation should mention the possible values

File: Sources/TuistKit/Commands/CleanCommand.swift (hunk 5 above)
Comment (fortmarek) [reply Sources/TuistKit/Commands/CleanCommand.swift:14]: While some are self-explanatory, for example `generatedAutomationProjects` could be a little confusing to the user. I think we should include the description of what the individual cache types do (probably both in the CLI and on the website).

Maybe even rewriting the autogenerated `default` bit might make sense here. I'd find `(default: whole tuist cache)` to be more readable than `(default: plugins, builds, tests, generatedAutomationProjects, projectDescriptionHelpers, manifests)`.

**Review by fortmarek — APPROVED** [review]

**Pull request merged**